facet-core/src/types/def/iter.rs

```diff
@@ -0,0 +1,152 @@
+
+/// VTable for an iterator
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+#[repr(C)]
+#[non_exhaustive]
+pub struct IterVTable<T: IterItem> {
+    /// cf. [`IterInitWithValueFn`]
+    pub init_with_value: Option<IterInitWithValueFn>,
+
+    /// cf. [`IterNextFn`]
+    pub next: IterNextFn<T>,
+
+    /// cf. [`IterNextBackFn`]
+    pub next_back: Option<IterNextBackFn<T>>,
+
+    /// cf. [`IterSizeHintFn`]
+    pub size_hint: Option<IterSizeHintFn>,
+
+    /// cf. [`IterDeallocFn`]
+    pub dealloc: IterDeallocFn,
+}
+
+impl<T: IterItem> IterVTable<T> {
+    /// Returns a builder for [`IterVTable`]
+    pub const fn builder() -> IterVTableBuilder<T> {
+        IterVTableBuilder::new()
+    }
+}
+
+/// Builds an [`IterVTable`]
+pub struct IterVTableBuilder<T: IterItem> {
+    init_with_value: Option<IterInitWithValueFn>,
+    next: Option<IterNextFn<T>>,
+    next_back: Option<IterNextBackFn<T>>,
+    size_hint: Option<IterSizeHintFn>,
+    dealloc: Option<IterDeallocFn>,
+}
+
+impl<T: IterItem> IterVTableBuilder<T> {
+    /// Creates a new [`IterVTableBuilder`] with all fields set to `None`.
+    #[allow(clippy::new_without_default)]
+    pub const fn new() -> Self {
+        Self {
+            init_with_value: None,
+            next: None,
+            next_back: None,
+            size_hint: None,
+            dealloc: None,
+        }
+    }
+
+    /// Sets the `init_with_value` function
+    pub const fn init_with_value(mut self, f: IterInitWithValueFn) -> Self {
+        self.init_with_value = Some(f);
+        self
+    }
+
+    /// Sets the `next` function
+    pub const fn next(mut self, f: IterNextFn<T>) -> Self {
+        self.next = Some(f);
+        self
+    }
+
+    /// Sets the `next_back` function
+    pub const fn next_back(mut self, f: IterNextBackFn<T>) -> Self {
+        self.next_back = Some(f);
+        self
+    }
+
+    /// Sets the `dealloc` function
+    pub const fn dealloc(mut self, f: IterDeallocFn) -> Self {
+        self.dealloc = Some(f);
+        self
+    }
+
+    /// Builds the [`IterVTable`] from the current state of the builder.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if any of the required fields are `None`.
+    pub const fn build(self) -> IterVTable<T> {
+        IterVTable {
+            init_with_value: self.init_with_value,
+            next: self.next.unwrap(),
+            next_back: self.next_back,
+            size_hint: self.size_hint,
+            dealloc: self.dealloc.unwrap(),
+        }
+    }
+}
+
+/// A kind of item that an [`IterVTable`] returns
+///
+/// This trait is needed as a utility, so the functions within [`IterVTable`]
+/// can apply the appropriate lifetime to their result types. In other words,
+/// this trait acts like a higher-kinded type that takes a lifetime.
+pub trait IterItem {
+    /// The output type of the iterator, bound by the lifetime `'a`
+    type Item<'a>;
+}
+
+impl IterItem for PtrConst<'_> {
```

Comment:
Also side note: there's actually no reason this needs to be implemented on `PtrConst<'_>` directly. The convention I've seen before would look like this:

```rust
// Empty enum (uninhabitable, only exists at compile-time)
pub enum PtrConst_ { }

impl IterItem for PtrConst_ {
  type Item<'a> = PtrConst<'a>;
}
```

It feels kinda cursed to impl it on the generic type directly, but it looks more intuitive